In this project, we will build a classification model of the Iris datatset in the Scikit-Learn library using the Random Forest Algorithm.

In [27]:
#import libraries
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification

In [28]:
#load the iris dataset
iris = datasets.load_iris()

In [6]:
#load the data
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

-After loading the data, we can see that there are 3 types of Iris (setosa, versicolor, virginica)
-There are 4 input features (sepal length, sepal width, petal length, petal width)(cm)
-There is 1 output variable (iris class label)

In [37]:
#create a dataframe for iris data set for visualization purposes
df = pd.DataFrame(data= np.c_[iris['data'], iris['target']],
                     columns= iris['feature_names'] + ['target'])

In [33]:
#print the first 5 rows
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0     0.0  
1     0.0  
2     0.0  
3     0.0  
4     0.0

In [7]:
#print input features (characteristics)
print(iris.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [8]:
#print output features (target names)
print(iris.target_names)

['setosa' 'versicolor' 'virginica']


In [9]:
#print output features in numerical value
print(iris.target)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [10]:
#print the data only
iris.data

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

To assign input and output variables, we have to assign the 4 input variables to X and the 1 output variable to Y.

In [12]:
X = iris.data
Y = iris.target

After assigning X and Y, we can examine the data dimension for both variables using the .shape function

In [13]:
X.shape

(150, 4)

This tells us that there are 150 rows which correspond to 150 flowers, and 4 columns which correspond to the input features (sepal length, sepal width, petal length, petal width)(cm).

In [14]:
Y.shape

(150,)

This tells us there are 150 rows corresponding to 150 flowers and there are no columns. 

Now that we have our X and Y in place, we can build a classification model using the Random Forest classification algorithm

In [15]:
#assign the Random Forest Classifier function as our classifier
clf = RandomForestClassifier()

In [16]:
#call the RFC using the .fit function to create a classification model using input arguments X (input features) & Y (class label)
clf.fit(X, Y)

RandomForestClassifier()

We will now look at the important features by using the .feature_impoprtances_ function which assigns the score of input features based on their importance to predict the output.

In [17]:
print(clf.feature_importances_)

[0.09566768 0.02294215 0.41993419 0.46145598]


The values are in order for the respectivce input features (sepal length, sepal width, petal length, petal width)(cm). This shows that the most important features here are petal length and petal width. With petal width being the most important, petal length being the 2nd most important, then sepal width, and lastly, sepal length. 

Now, we can start making predictions.

In [19]:
#call the 2nd flower
X[1]

array([4.9, 3. , 1.4, 0.2])

We will now use the .predict function from our classifier using the set of values of the 2nd flower to predict the class of the 2nd flower in the data set.

In [46]:
#allows us to predict the class of the flower using the class name(target names) instead of values(target) ex: (0,1,2)
clf.fit(iris.data, iris.target_names[iris.target])

RandomForestClassifier()

In [47]:
print(clf.predict(X[[1]]))

['setosa']


Now we can use the .predict_proba() method which returns an array containing the class probabilities of the input data.

In [42]:
print(clf.predict_proba(X[[1]]))

[[1. 0. 0.]]


Above we can see that the 2nd flower from the Iris data set has 100% prbabiity of being an Iris-Setosa.

Now to train and test our data, we will be splitting the data as an 80/20 ratio. Where 80% of the data will go to the training set and 20% will go to the testing set. We will do this by using the train_test_split function we imported fom scikit learn.

In [48]:
#Assign X train/test, and Y train/test to train_test_split function, giving the test set a value of 0.2 or 20% of the data from the data set
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2)

We can use the .shape function to see how many columns there are in our training set and in the test set with the given 80/20 ratio.

In [49]:
X_train.shape, Y_train.shape

((120, 4), (120,))

In [50]:
X_test.shape, Y_test.shape

((30, 4), (30,))

After splitting the data into our training and testing set, we have to rebuild the random forest model using X_train and Y_train as our input arguments. 

In [51]:
clf.fit(X_train, Y_train)

RandomForestClassifier()

Now we can perform a prediction of a single sample from the data set, in this case, we will be using the 3rd flower from the data set, or X[2]. Using the .predict function, the model will predict the class and using the .predict_proba, the model will return the probability of the prediction being correct. 

In [52]:
print(clf.predict(X[[2]]))
print(clf.predict_proba(X[[2]]))

[0]
[[1. 0. 0.]]


Instead of only using single samples from our data set to make a prediction, we can use a larger sample such as our X_test to predict the class label. 

In [53]:
#predictive class labels
print(clf.predict(X_test))

[2 2 2 2 2 1 0 0 0 1 2 0 1 0 0 0 1 2 2 1 1 2 0 2 1 1 0 0 1 2]


Now we can print the Y_test which shows the actual class labels. 

In [55]:
#actual class labels
print(Y_test)

[2 2 2 2 2 1 0 0 0 1 1 0 1 0 0 0 1 2 2 1 1 2 0 1 1 1 0 0 1 2]


By comparing the class labels, we can see that there are a few differences, we can use the .score function using X_test and Y_test as input arguments which returns the mean accuracy on the given test data and labels.

In [57]:
print(clf.score(X_test, Y_test))

0.9333333333333333
